README.md
# qbat - Light qt5 tool to display battery stats in system tray.


debian/changelog
qbat (0.2.2astra18) unstable; urgency=medium
  * try to get qbat apps icon from theme
 -- Alexander Smirnov <support@rusbitech.ru>  Tue, 24 Oct 2017 19:52:40 +0300
qbat (0.2.2astra17) unstable; urgency=medium
  * merge batteries by default
  * create battery icons in spite of /sys/class/power_supply/.../online file
 -- Alexander Smirnov <support@rusbitech.ru>  Thu, 05 Oct 2017 17:31:31 +0300
qbat (0.2.2astra16) unstable; urgency=medium
  * use fly help simple  API
 -- Alexander Smirnov <support@rusbitech.ru>  Thu, 25 May 2017 15:27:26 +0300
qbat (0.2.2astra15) unstable; urgency=medium
  * rebuild
 -- Alexander Smirnov <support@rusbitech.ru>  Thu, 23 Mar 2017 20:46:11 +0300
qbat (0.2.2-astra14) unstable; urgency=low
  * rules fix
 -- Alexander Smirnov <support@rusbitech.ru>  Fri, 06 May 2016 18:15:32 +0300
qbat (0.2.2-astra13) unstable; urgency=low
  * second call qbat => show tooltip
 -- Alexander Smirnov <support@rusbitech.ru>  Mon, 21 Sep 2015 18:45:04 +0300
qbat (0.2.2-astra12) unstable; urgency=low
  * sony tablet z, kernel 3.4: check if rate > 0
  * sony tablet z, kernel 3.4: check if get 'capacity'<=100 use it, dont use wrong charge_now
 -- Alexander Smirnov <support@rusbitech.ru>  Fri, 05 Jun 2015 18:45:04 +0300
qbat (0.2.2-astra10) unstable; urgency=low
  * check if can't get fullCapacity from charge_full file,then currentCapacity (charge_now) may be in percents
 -- Alexander Smirnov <support@rusbitech.ru>  Fri, 03 Apr 2015 18:45:04 +0300
qbat (0.2.2-astra9) unstable; urgency=low
  * option to use standard theme icons if available
  * resize painted tray icon to 64x64
 -- Alexander Smirnov <support@rusbitech.ru>  Wed, 18 Mar 2015 18:45:04 +0300
qbat (0.2.2-astra8) unstable; urgency=low
  * build fix
 -- Alexander Smirnov <support@rusbitech.ru>  Wed,  2 Jul 2014 17:45:04 +0400
qbat (0.2.2-astra7) unstable; urgency=low
  * fly help
 -- Alexander Smirnov <support@rusbitech.ru>  Thu, 19 Jun 2014 17:45:04 +0400
qbat (0.2.2-astra6) unstable; urgency=low
  * libqt5x11extras5-dev
 -- Alexander Smirnov <support@rusbitech.ru>  Thu, 06 Mar 2014 17:45:04 +0400
qbat (0.2.2-astra5) unstable; urgency=low
  * buildfix
  * new public API flydeapi.h
 -- Alexander Smirnov <support@rusbitech.ru>  Tue, 28 Jan 2014 15:49:28 +0400
qbat (0.2.2-astra4) unstable; urgency=low
  * Port to Qt 5
 -- Alexander Volkov <support@rusbitech.ru>  Fri, 13 Dec 2013 14:14:55 +0400
qbat (0.2.2-astra3) unstable; urgency=low
  * buildfix
 -- Alexander Smirnov <rusbitech@mail.ru>  Tue, 26 Feb 2013 18:03:45 +0300
qbat (0.2.2-astra2) unstable; urgency=low
  * buildfix
 -- Alexander Smirnov <rusbitech@mail.ru>  Thu, 31 Jan 2013 18:03:45 +0300
qbat (0.2.2-astra1) unstable; urgency=low
  * Comment added to desktop file
 -- Alexander Smirnov <rusbitech@mail.ru>  Thu, 11 Oct 2012 18:03:45 +0300
qbat (0.2.2-astra0) unstable; urgency=low
  * sync with original
 -- Alexander Smirnov <rusbitech@mail.ru>  Tue, 10 Jul 2012 18:03:45 +0300
qbat (0.1.8-astra7) unstable; urgency=low
  * session management stuff
 -- Alexander Smirnov <sja@mail.ru>  Sat,  7 May 2009 18:03:45 +0300
qbat (0.1.8-astra6) unstable; urgency=low
  * fix autostart
 -- Alexander Smirnov <sja@mail.ru>  Thu, 24 Sep 2009 18:03:45 +0300
qbat (0.1.8-astra5) unstable; urgency=low
  * dependencies fixed
 -- Alexander Smirnov <sja@mail.ru>  Thu, 23 Jul 2009 18:03:45 +0300
qbat (0.1.8-astra4) unstable; urgency=low
  * translation enhanced
 -- Alexander Smirnov <sja@mail.ru>  Mon, 15 May 2009 18:03:45 +0300
qbat (0.1.8-astra3) unstable; urgency=low
  * signle application check
 -- Alexander Smirnov <sja@mail.ru>  Wed,  6 May 2009 18:03:45 +0300
qbat (0.1.8-astra2) unstable; urgency=low
  * translation fixes
  * autostart option
 -- Alexander Smirnov <sja@mail.ru>  Wed, 22 Apr 2009 14:03:45 +0300
qbat (0.1.8-astra1) unstable; urgency=low
  * first astra-build with minor fixes
 -- Alexander Smirnov <sja@mail.ru>  Mon, 20 Apr 2009 20:03:45 +0300


qbat (0.2.2) unstable; urgency=low

  * fixed battery loss information
  * settings are now saved by accepting the settings dialog

 -- Oliver Groß <neokensuke@freenet.de>  Mon, 20 Jul 2009 14:11:05 +0200

qbat (0.2.1) unstable; urgency=low

  * .desktop file fixes
  * added dis-/charging color settings
  * left click on (battery-)icon shows current information
  * added capacity loss information

 -- Oliver Groß <neokensuke@freenet.de>  Wed, 15 Jul 2009 12:15:50 +0200

qbat (0.2.0) unstable; urgency=low

  * split timers (for lower CPU usage):
    - check for power supplies
    - retrieve power supply data
  * new icons
  * code cleanup
  * internal redesign
  * german translation

 -- Oliver Groß <neokensuke@freenet.de>  Thu, 18 Jun 2009 13:08:41 +0200

qbat (0.1.8) unstable; urgency=low

  * code cleanup
  * display current rate in watt and ampere
  * fixed repeated critical event handling
  * workaround for delayed update of battery icon colors

 -- Oliver Groß <neokensuke@freenet.de>  Tue, 07 Oct 2008 16:03:45 +0200

qbat (0.1.7) unstable; urgency=low

  * fixed "command<->waring" setting
  * fixed sysfs filenames
  * fixed lintian copyright warning

 -- Oliver Groß <neokensuke@freenet.de>  Tue, 02 Sep 2008 22:11:25 +0200

qbat (0.1.6) unstable; urgency=low

  * added basic support for battery stats in mWh
  * added confirmation dialog before executing command
    (with optional timeout)
  * added gentoo ebuild
  * fixed setting command

 -- Oliver Groß <neokensuke@freenet.de>  Tue, 02 Sep 2008 20:21:16 +0200

qbat (0.1.5) unstable; urgency=low

  * added option to display balloon tips instead of dialog messages
  * added charging time information
  * removed obsolete d-bus dependecy
  * fixed .desktop file
  * some Qt 4.4.x fixes

 -- Oliver Groß <neokensuke@freenet.de>  Mon, 02 Jun 2008 15:52:59 +0200

qbat (0.1.4) unstable; urgency=low

  * Fixed and improved battery icon painting
  * Added caching of information

 -- Oliver Groß <neokensuke@freenet.de>  Thu, 17 Apr 2008 14:13:30 +0200

qbat (0.1.3) unstable; urgency=low


batteryicon.cpp
//
// C++ Implementation: batteryicon
//
// Author: Oliver Groß <z.o.gross@gmx.de>, (C) 2008
//
// Copyright: See COPYING file that comes with this distribution
//
#include "batteryicon.h"
#include "common.h"
#include <QPainter>
#include <QMessageBox>

namespace qbat {
	QDir CBatteryIcon::sysfsDir(UI_PATH_SYSFS_DIR);
	
	CBatteryIcon::CBatteryIcon(Settings * settings, QString batteryName, QObject * parent) :
		QSystemTrayIcon(parent),
		m_Icon(28, 28),
		m_IconName(""), //alex
		m_Settings(settings)
	{
		m_Data.name = batteryName;
		
		updateIcon();
		
		connect(this, SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
			this, SLOT(handleClicks(QSystemTrayIcon::ActivationReason)));
	}
	
	CBatteryIcon::~CBatteryIcon() {
	}
	
	void CBatteryIcon::updateIcon() {
	    if (m_Settings->standardIcons) //alex
	    {
	        //name is "battery[-charging][-percent]", where percent may be "caution",low,40,60,80,100 or ""
		QString m_IconName, charging, percent;

		if (m_Data.status == UI_BATTERY_CHARGING) charging="-charging";
		if (m_Data.relativeCharge <= 100)
		{
			 if (m_Data.relativeCharge< 0) percent="-caution";//-1 is smth. wrong?
		    else if (m_Data.relativeCharge<20) percent="-low";  //00-20
		    else if (m_Data.relativeCharge<40) percent="-040";  //20-40
		    else if (m_Data.relativeCharge<60) percent="-060";  //40-60
		    else if (m_Data.relativeCharge<80) percent="-080";  //60-80
		    else if (m_Data.relativeCharge<=100) 
		      percent=((m_Data.status==UI_BATTERY_DISCHARGING || m_Data.status==UI_BATTERY_FULL)?"-100":""); //80-100
		}
		else
		{
		  //>100 is the same as 100?
		  if (m_Data.relativeCharge<=100) 
		    percent=((m_Data.status==UI_BATTERY_DISCHARGING || m_Data.status==UI_BATTERY_FULL)?"-100":"");
		}
		m_IconName = QString("battery")+charging+percent;
		QIcon ico = QIcon::fromTheme(m_IconName);
		if (ico.isNull()) goto draw_it;
		setIcon(ico);
		//fprintf(stderr,"set icon=%s rel=%d\n",m_IconName.toLocal8Bit().data(),m_Data.relativeCharge);
	    }
	    else
	    {
draw_it:
		m_Icon.fill(Qt::transparent);
		QPainter painter(&m_Icon);
		
		if (m_Data.currentCapacity != m_Data.fullCapacity || m_Data.status != UI_BATTERY_FULL) {
			if (m_Data.status == UI_BATTERY_DISCHARGING)
				painter.setPen(QColor(m_Settings->colors[UI_COLOR_PEN_DISCHARGE]));
			else
				painter.setPen(QColor(m_Settings->colors[UI_COLOR_PEN_CHARGE]));
			
			painter.setBrush(QColor(m_Settings->colors[UI_COLOR_BRUSH_EMPTY]));
			painter.drawRect(0, 4, 27, 23);
			
			int chargedPixels = (int)(22 * m_Data.relativeCharge / 100.0);
			
			painter.fillRect(1, 5 + 22 - chargedPixels, 26, chargedPixels, QColor(m_Settings->colors[UI_COLOR_BRUSH_CHARGED]));
			
			if (m_Data.status == UI_BATTERY_DISCHARGING)
				painter.setBrush(QColor(m_Settings->colors[UI_COLOR_BRUSH_POLE_DISCHARGE]));
			else
				painter.setBrush(QColor(m_Settings->colors[UI_COLOR_BRUSH_POLE_CHARGE]));
		}
		else {
			painter.setPen(QColor(m_Settings->colors[UI_COLOR_PEN_FULL]));
			painter.setBrush(QColor(m_Settings->colors[UI_COLOR_BRUSH_FULL]));
			painter.drawRect(0, 4, 27, 23);
			
			painter.setBrush(QColor(m_Settings->colors[UI_COLOR_BRUSH_POLE_FULL]));
		}
		painter.drawRect(7, 0, 13, 4);
		
		painter.setBrush(Qt::NoBrush);
		
		if (m_Data.relativeCharge < 100)
			((QFont&)painter.font()).setPixelSize(15);
		else
			((QFont&)painter.font()).setPixelSize(12);
		
		painter.setRenderHint(QPainter::TextAntialiasing);
		((QFont&)painter.font()).setBold(true);
		if (m_Data.relativeCharge == -1)
			painter.drawText(1, 9, 26, 16, Qt::AlignHCenter, QString('?'));
		else
			painter.drawText(1, 9, 26, 16, Qt::AlignHCenter, QString::number(m_Data.relativeCharge));
		
		setIcon(m_Icon.scaled(QSize(128,128))); //alex
	    }
	}
	
	void CBatteryIcon::updateToolTip() {
		QString newToolTip = tr("%1: %2%").arg(m_Data.name) +'\n';
		
		if (m_Data.relativeCharge == -1)
			newToolTip = newToolTip.arg('-');
		else
			newToolTip = newToolTip.arg(m_Data.relativeCharge);
		
		switch (m_Data.status) {
		case UI_BATTERY_DISCHARGING:
			newToolTip += tr("status: %1").arg(tr("discharging"));
			if (m_Data.rate>0 && m_Data.fullCapacity>100) { //alex: if fullCap not in percents
				newToolTip += '\n';
				qreal remainingTime  = (qreal)(m_Data.currentCapacity) / m_Data.rate;
				int remainingHours   = (int)remainingTime;
				int remainungMinutes = (int)(remainingTime * 60) % 60;
				newToolTip += tr("remaining time: %1:%2").arg(remainingHours, 2, 10, QChar('0')).arg(remainungMinutes, 2, 10, QChar('0'));
			}
			break;
		case UI_BATTERY_CHARGING:
			newToolTip += tr("status: %1").arg(tr("charging"));
			if (m_Data.rate>0 && m_Data.fullCapacity>100) { //alex: if fullCap not in percents
				newToolTip += '\n';
				qreal remainingTime  = (qreal)(m_Data.fullCapacity - m_Data.currentCapacity) / m_Data.rate;
				int remainingHours   = (int)remainingTime;
				int remainungMinutes = (int)(remainingTime * 60) % 60;
				newToolTip += tr("remaining time: %1:%2").arg(remainingHours, 2, 10, QChar('0')).arg(remainungMinutes, 2, 10, QChar('0'));
			}
			break;
		case UI_BATTERY_FULL:
			newToolTip += tr("status: %1").arg(tr("full"));
			break;
		default:
			newToolTip += tr("status: %1").arg(tr("unknown"));
			break;
		}
		newToolTip += '\n';
		
		if (m_Data.energyUnits) {
			if ((m_Data.rate>0) && (m_Data.status != UI_BATTERY_FULL)) {
				double rateW = qRound(m_Data.rate / 100000.0) / 10.0;
				if (m_Data.voltage) {
					double rateA = qRound((m_Data.rate / m_Data.voltage) / 1000.0) / 10.0;
					newToolTip += tr("current rate: %1W / %2A").arg(rateW).arg(rateA) + '\n';
				}
				else
					newToolTip += tr("current rate: %1W").arg(rateW) + '\n';
			}
			
			if (m_Data.fullCapacity>100) //alex: if fullCap not in percents
			newToolTip += tr("current capacity: %1mWh").arg(m_Data.currentCapacity / 1000);
			
			if (m_Data.fullCapacity>100)
				newToolTip += '\n' + tr("last full capacity: %1mWh").arg(m_Data.fullCapacity / 1000);
			
			if (m_Data.designCapacity)
				newToolTip += '\n' + tr("design capacity: %1mWh").arg(m_Data.designCapacity / 1000);
		}
		else
		{
			if ((m_Data.rate>0) && (m_Data.status != UI_BATTERY_FULL)) {
				double rateA = m_Data.rate / 100000.0;
				if (m_Data.voltage) {
					double rateW = qRound(rateA * m_Data.voltage / 100.0) / 10.0;
					if (rateW>0 && rateA>0) //alex
					newToolTip += tr("current rate: %1W / %2A").arg(rateW).arg(qRound(rateA) / 10.0) + '\n';
				}
				else {
					if (rateA>0) //alex
					newToolTip += tr("current rate: %1A").arg(qRound(rateA) / 10.0) + '\n';
				}
			}
			
			if (m_Data.fullCapacity>100) //alex: if fullCap not in percents
			newToolTip += tr("current capacity: %1mAh").arg(m_Data.currentCapacity / 1000);
			
			if (m_Data.fullCapacity>100) //alex: if fullCap not in percents
				newToolTip += '\n' + tr("last full capacity: %1mAh").arg(m_Data.fullCapacity / 1000);
			
			if (m_Data.designCapacity)
				newToolTip += '\n' + tr("design capacity: %1mAh").arg(m_Data.designCapacity / 1000);
		}
		
		if (m_Data.designCapacity && m_Data.fullCapacity < m_Data.designCapacity)
				newToolTip += '\n' + tr("capacity loss: %1%").arg((int)((1 - (double)m_Data.fullCapacity / m_Data.designCapacity) * 100));
		
		setMessage(newToolTip);
	}
	
	void CBatteryIcon::updateData(int currentCapacity, int fullCapacity, int designCapacity, int rate, int voltage, int status, bool energyUnits) {
		m_Data.energyUnits = energyUnits;
		
		bool statusUpdated=false; //alex
		
		bool noupdate = true;
		
		if (rate != m_Data.rate) {
			noupdate = false;
			m_Data.rate = rate;
		}
		
		if (voltage != m_Data.voltage) {
			noupdate = false;
			m_Data.voltage = voltage;
		}
		
		if (status != m_Data.status) {
			noupdate = false;
			m_Data.status = status;
			statusUpdated=true; //alex
		}
		
		if (fullCapacity != m_Data.fullCapacity) {
			noupdate = false;
			m_Data.fullCapacity = fullCapacity;
		}
		
		if (designCapacity != m_Data.designCapacity) {
			noupdate = false;
			m_Data.designCapacity = designCapacity;
		}
		
		if (currentCapacity != m_Data.currentCapacity) {
			noupdate = false;
			m_Data.currentCapacity = currentCapacity;
		}
		
		if (noupdate)
			return;
		
		qint8 newRelativeCharge = calcRelativeDef(currentCapacity, fullCapacity);
		
		if (newRelativeCharge != m_Data.relativeCharge || statusUpdated) { //alex: add status check
			m_Data.relativeCharge = newRelativeCharge;
			
			if (isVisible())
				updateIcon();
		}
		
		if (isVisible())
			updateToolTip();
	}
	
	void CBatteryIcon::updateData() {
		int currentCapacity = 0;
		int fullCapacity = 0;
		int designCapacity = 0;
		int rate = 0;
		int voltage = 0;
		int status = 0;
		bool energyUnits = sysfsDir.exists(m_Data.name + UI_CAPTION_NOW(UI_CAPTION_ENERGY));
		
		rate = readIntSysFile(sysfsDir.filePath(m_Data.name + "/current_now").toLatin1().constData());
		voltage = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_NOW(UI_CAPTION_VOLTAGE)).toLatin1().constData()) / 10000;
		
		char buffer[BUF_SIZE];
		readStringFromFile(buffer, sysfsDir.filePath(m_Data.name + "/status").toLatin1().constData());
		status = toStatusInt(buffer);
		
		if (energyUnits) {
			fullCapacity    = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_FULL(UI_CAPTION_ENERGY)).toLatin1().constData());
			designCapacity  = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_DESIGN(UI_CAPTION_ENERGY)).toLatin1().constData());
			currentCapacity = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_NOW(UI_CAPTION_ENERGY)).toLatin1().constData());
		}
		else {
			fullCapacity    = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_FULL(UI_CAPTION_CHARGE)).toLatin1().constData());
			designCapacity  = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_DESIGN(UI_CAPTION_CHARGE)).toLatin1().constData());
			currentCapacity = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_NOW(UI_CAPTION_CHARGE)).toLatin1().constData());
			int capacityProbe  = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_CAPACITY).toLatin1().constData());
			if (fullCapacity==0 ) //alex
			{
			  currentCapacity = readIntSysFile(sysfsDir.filePath(m_Data.name + UI_CAPTION_CAPACITY).toLatin1().constData());
			  fullCapacity=100;
			}
			else if (capacityProbe<=100) //alex: kernel 3.4 on sony tablet z: 'capacity' field in percents, current capacity iws smth wrong
			{
			  currentCapacity =  fullCapacity * capacityProbe / 100;
			}
		}
		updateData(currentCapacity, fullCapacity, designCapacity, rate, voltage, status, energyUnits);
	}
	
	void CBatteryIcon::setMessage(QString value) {
		m_Message = value;
		setToolTip(tr("QBat - %1").arg(m_Message));
	}
	
	void CBatteryIcon::handleClicks(QSystemTrayIcon::ActivationReason reason) {
		switch (reason) {
			case Trigger:
				if (supportsMessages())
					showMessage(tr("QBat - Information"), m_Message);
				else
					QMessageBox::information(NULL, tr("QBat - Information"), m_Message);
				break;
			default:
				break;
		}
	}
}
